internal/keepers/util.go

```diff
@@ -110,88 +111,146 @@ func dedupe[T fmt.Stringer](inputs [][]T, filters ...func(T) bool) ([]T, error)
 	return output, nil
 }
 
-func shuffledDedupedKeyList(attributed []types.AttributedObservation, key [16]byte, filters ...func(ktypes.UpkeepKey) bool) ([]ktypes.UpkeepKey, error) {
-	var err error
+func shuffleDedupedObservations(upkeepKeys [][]ktypes.UpkeepKey, key [16]byte, filters ...func(ktypes.UpkeepKey) bool) ([]ktypes.UpkeepKey, error) {
+	uniqueKeys, err := dedupe(upkeepKeys, filters...)
+	if err != nil {
+		return nil, err
+	}
 
-	if len(attributed) == 0 {
-		return nil, fmt.Errorf("%w: must provide at least 1", ErrNotEnoughInputs)
+	uniqueKeys, err = trimLowerBlocks(uniqueKeys)
+	if err != nil {
+		return nil, err
 	}
 
-	var parseErrors int
-	kys := make([][]ktypes.UpkeepKey, len(attributed))
-	for i, attr := range attributed {
-		b := []byte(attr.Observation)
-		if len(b) == 0 {
+	rnd.New(util.NewKeyedCryptoRandSource(key)).Shuffle(len(uniqueKeys), func(i, j int) {
+		uniqueKeys[i], uniqueKeys[j] = uniqueKeys[j], uniqueKeys[i]
+	})
+
+	return uniqueKeys, nil
+}
+
+func trimLowerBlocks(uniqueKeys []ktypes.UpkeepKey) ([]ktypes.UpkeepKey, error) {
+	idxMap := make(map[string]int)
+	out := make([]ktypes.UpkeepKey, 0, len(uniqueKeys))
+
+	for _, uniqueKey := range uniqueKeys {
+		blockKey, upkeepID, err := uniqueKey.BlockKeyAndUpkeepID()
+		if err != nil {
+			return nil, err
+		}
+
+		idx, ok := idxMap[string(upkeepID)]
+		if !ok {
+			idxMap[string(upkeepID)] = len(out)
+			out = append(out, uniqueKey)
 			continue
 		}
 
-		var ob []ktypes.UpkeepKey
+		savedBlockKey, _, err := out[idx].BlockKeyAndUpkeepID()
+		if err != nil {
+			return nil, err
+		}
+
+		if string(blockKey) > string(savedBlockKey) {
+			out[idx] = uniqueKey
+		}
+	}
+
+	return out, nil
+}
 
+func observationsToUpkeepKeys(observations []types.AttributedObservation, reportBlockLag int) ([][]ktypes.UpkeepKey, error) {
+	var parseErrors int
+
+	res := make([][]ktypes.UpkeepKey, len(observations))
+
+	var allBlockKeys []ktypes.BlockKey
+	for i, observation := range observations {
 		// a single observation returning an error here can void all other
 		// good observations. ensure this loop continues on error, but collect
 		// them and throw an error if ALL observations fail at this point.
 		// TODO we can't rely on this concrete type for decoding/encoding
-		var keys []chain.UpkeepKey
-		err = decode(b, &keys)
-		if err != nil {
+		var upkeepObservation *ktypes.UpkeepObservation
+		if err := decode(observation.Observation, &upkeepObservation); err != nil {
+			fmt.Println("err", err)
 			parseErrors++
 			continue
 		}
 
-		for _, o := range keys {
-			ob = append(ob, o)
-		}
+		allBlockKeys = append(allBlockKeys, upkeepObservation.BlockKey)
 
-		sort.Sort(sortUpkeepKeys(ob))
-		kys[i] = ob
+		// if we have a non-empty list of upkeep identifiers, use the zeroth upkeep identifier
+		if len(upkeepObservation.UpkeepIdentifiers) > 0 {
+			res[i] = []ktypes.UpkeepKey{
+				chain.NewUpkeepKeyFromBlockAndID(upkeepObservation.BlockKey, upkeepObservation.UpkeepIdentifiers[0]),
```

Comment:
this is hardcoded to be 0. instead we should pass `observationUpkeepsLimit` to this function as use that to limit slice

Reply:
also not sure why we are appending block number here. we can just keep list of upkeep identifiers as we append the block number later anyway